packages/context/src/binding-filter.ts

```diff
@@ -9,6 +9,15 @@ import {BindingAddress} from './binding-key';
 /**
  * A function that filters bindings. It returns `true` to select a given
  * binding.
+ *
+ * Note: originally, we allow filters to be tied with a single value type.
+ * This actually does not make much sense - the filter function is typically
+ * invoked on all bindings to find those matching the given criteria,
+ * therefore filters must be prepared to handle bindings of any value type.
+ * We learned about this problem after enabling TypeScript's `strictFunctionTypes`
+ * check, but decided to preserve `ValueType` argument for backwards compatibility.
+ *
+ * TODO(semver-major): remove ValueType template argument.
```

Comment:
See https://github.com/strongloop/loopback-next/pull/2728. The `ValueType` is used internally with a default. Applications or extensions may define their own filter functions and mostly won't cast it to `BindingFilter<...>`. I don't think we have to release a new major.